README

Coursera Downloader
-------------------

Coursera (http://coursera.org) is creating some fantastic, free educational
videos (e.g., algorithms, machine learning, natural language processing, SaaS).
This script allows one to batch download videos for a Coursera class.  Given a
class name and related cookie file, it scrapes the course listing page to get
the week and class names, and then downloads the related videos into
appropriately named files and directories.  

Why is this helpful?  Before I was using wget, but I had the following problems:
  1. Video names have a number in them, but this does not correspond to the
     actual order.  Manually renaming them is a pain.
  2. Using names from the syllabus page provides more informative names.
  3. Using a wget in a forloop picks up extra videos which are not
     posted/linked, and these are sometimes duplicates.

DownloadThemAll can also work, but this provides better names.

Features:
  * Intentionally detailed names, so that it will display and sort properly
    using MX Video on my Andriod phone.
  * Regex-based section (week) and video name filters to download only certain
    resources.
  * Tested on both Linux and Windows.

Inspired in part by youtube-dl (http://rg3.github.com/youtube-dl) by which I've
downloaded many other good videos such those from Khan Academy.  


Directions
----------

Requires Python 2.x (where x >= 5) and a free Coursera account.

1. Install any missing dependencies.

  1a. Beautiful Soup 3 -- http://www.crummy.com/software/BeautifulSoup/
    Ubuntu/Debian: sudo apt-get install python-beautifulsoup 
    Other*: easy_install BeautifulSoup

  1b. Argparse -- http://pypi.python.org/pypi/argparse
    (Not necessary if Python version >= 2.7)
    Ubuntu/Debian: sudo apt-get install python-argparse
    Other*: easy_install argparse

  * To install "easy_install"
    Ubuntu: sudo apt-get install python-setuptools
    Windows: http://pypi.python.org/pypi/setuptools

  1c. wget (needed on Windows)
    http://sourceforge.net/projects/gnuwin32/files/wget/1.11.4-1/wget-1.11.4-1-setup.exe/
    This is a temporary workaround as currently the python download 
    has a problem.

2. Create a Coursera.org account and login.
    e.g. http://saas-class.org

3. Locate or export your Netscape-style cookies file with a browser extension.
    Chrome: Cookie.txt Export
      https://chrome.google.com/webstore/detail/lopabhfecdfhgogdbojmaicoicjekelh
    Firefox: Export Cookies 1.2
      https://addons.mozilla.org/en-US/firefox/addon/export-cookies/

4. Run the script to download the videos.
    Linux:
      coursera-dl saas -c cookies.txt
    # filter by section name
      coursera-dl saas -c cookies.txt -sf "Chapter_Four"
    # filter by video name
      coursera-dl saas -c cookies.txt -vf "3.1_"
    Windows:
      You need to use wget right now.
      coursera-dl saas -c cookies.txt -w C:\...\wget.exe


Troubleshooting
---------------

If it's finding 0 sections, you probably have a bad cookies file.
Use -l listing.html and then examine that file -- if it's the non-logged-in
page then this is definitely your problem.

Tested on:
  - Ubuntu/Python 2.6.5
  - Win7/Python2.5 (with wget)


Contact
-------
Send any questions, comments, or problems to:
  John Lehmann: first last at geemail dotcom or @jplehmann


CHANGELOG.md
# Change Log

## 0.11.5 (2019-12-16)

Features:

-   add --cauth argument to specify CAUTH cookie directly from command-line (#724)

## 0.11.4 (2018-06-24)

Features:

-   Do not expand class names if there is a specialization with the same name,
    but add --specialization flag to do that explicitly (#673)

## 0.11.3 (2018-06-24)

Bugfixes:

-   Switch to newer API for syllabus and lecture retrieval (#665, #673, #634)

Features:

-   You can now download specializations: the child courses will be
    downloaded automatically

## 0.11.2 (2018-06-03)

Bugfixes:

-   Use TLS v1.2 instead of v1.0
-   Switched to api.coursera.org subdomain for subtitles requests (#664)

## 0.11.1 (2018-06-02)

Bugfixes:

-   Specify utf-8 encoding in setup.py to fix installation on Windows (#662)

## 0.11.0 (2018-06-02)

Features:

-   Add support for "peer assignment" section (#650)

Bugfixes:

-   Switched to api.coursera.org subdomain for API requests (#660)

## 0.10.0 (2018-02-19)

Features:

-   Support Coursera Notebooks (option: `--download-notebooks`)
-   Add hints in the documentation for users in China

## 0.9.0 (2017-05-25)

Features:

-   Default arguments are loaded from `coursera-dl.conf` file
-   Added option `--mathjax-cdn <MATHJAX_CDN>` to specify alternative MathJax CDN
-   Added support for Resources section

## 0.8.0 (2016-10-04)

Features:

-   Add `--download-delay` option that adds a specified delay in seconds
    before downloading next course. This is useful when downloading many
    courses at once. Default value is 60 seconds.
-   Add `--only-syllabus` option which is when activated, allows to skip
    download of the course content. Only syllabus is parsed.
-   Add support for `reflect` and `mcqReflect` question types in quizzes.
-   Courses that encountered an error while parsing syllabus will be listed
    in the end of the program execution, after all courses have been
    processed (hopefully, downloaded). This helps skip vast output and easily
    see which courses need user's attention, e.g. enrollment, session
    switching or just patience until the course start date.

Bugfixes:

-   Locked programming assignments in syllabus used to crash coursera-dl.
    Now the script goes on parsing syllabus and skips locked assignments.
-   Add missing import statement to playlist generation module

## 0.7.0 (2016-07-28)

Features:

-   Added option `--list-courses` to list currently enrolled courses (#514)
-   Added option `--jobs N` to download resources in N threads simultaneously (#553)
-   Added option `--download-quizzes` to download contents of graded and
    ungraded quizzes (#490)
-   Added option `--cache-syllabus` to avoid downloading course syllabus on
    every run (this option is rather for developers)

Bugfixes:

-   Locked lectures are also requested from server now, which allows to
    download some programming assignments that were not downloaded before (#555)

Deletions:

-   Support for old-style courses has been removed (Coursera discontinued old courses:
    http://coursera.tumblr.com/post/145882467032/courseras-transition-to-a-new-technology-platform)
-   `--ignore-http-errors` option has been removed and the default behavior
    has been adjusted to include this option
-   Removed deprecated `--on-demand` option. Now OnDemand classes are downloaded
    by default

## 0.6.1 (2016-06-20)

Bugfixes:

-   When using `--process_local_page` option, errors downloading About
    page will not stop course download
-   Limit file name part to 200 characters and file extension part to 20
    characters to alleviate "Filename is too long" issue

## 0.6.0 (2016-06-17)

Features:

-   Descriptions of assignments are saved in more cases (different courses
    implement it in a different way)
-   Images are embedded into descriptions of assignments and are available
    offline locally
-   MathJax is injected into descriptions of assignments to render math
    locally (requires Internet connection)
-   Add option `--ignore-http-errors` to ignore errors in internal
    downloader during resource downloading stage
-   Add option `--disable-url-skipping` to always try downloading
    all the URLs that were found by the parser
-   Add option `--downloader-arguments` (#464)
-   Add option `--version` (#477)
-   Better looking progress output: during the syllabus parsing stage
    lecture/section names are printed in a hierarchical structure

Bugfixes:

-   Stricter filename cleaning in on-demand course parser
-   Better URL filtering
-   Detect SSL errors and suggest link to the solution in the output
-   Added workaround for "Filename is too long" on Windows


coursera/cookies.py
# -*- coding: utf-8 -*-

"""
Cookie handling module.
"""

import logging
import os
import ssl

import requests
from requests.adapters import HTTPAdapter

try:  # Workaround for broken Debian/Ubuntu packages? (See issue #331)
    from requests.packages.urllib3.poolmanager import PoolManager
except ImportError:
    from urllib3.poolmanager import PoolManager

from six.moves import StringIO
from six.moves import http_cookiejar as cookielib

from .define import AUTH_REDIRECT_URL, AUTH_URL_V3, CLASS_URL, PATH_COOKIES
from .utils import mkdir_p, random_string

# Monkey patch cookielib.Cookie.__init__.
# Reason: The expires value may be a decimal string,
# but the Cookie class uses int() ...
__original_init__ = cookielib.Cookie.__init__


def __fixed_init__(
    self,
    version,
    name,
    value,
    port,
    port_specified,
    domain,
    domain_specified,
    domain_initial_dot,
    path,
    path_specified,
    secure,
    expires,
    discard,
    comment,
    comment_url,
    rest,
    rfc2109=False,
):
    if expires is not None:
        expires = float(expires)
    __original_init__(
        self,
        version,
        name,
        value,
        port,
        port_specified,
        domain,
        domain_specified,
        domain_initial_dot,
        path,
        path_specified,
        secure,
        expires,
        discard,
        comment,
        comment_url,
        rest,
        rfc2109=False,
    )


cookielib.Cookie.__init__ = __fixed_init__


class ClassNotFound(BaseException):
    """
    Raised if a course is not found in Coursera's site.
    """


class AuthenticationFailed(BaseException):
    """
    Raised if we cannot authenticate on Coursera's site.
    """


def prepare_auth_headers(session, include_cauth=False):
    """
    This function prepares headers with CSRF/CAUTH tokens that can
    be used in POST requests such as login/get_quiz.

    @param session: Requests session.
    @type session: requests.Session

    @param include_cauth: Flag that indicates whether CAUTH cookies should be
        included as well.
    @type include_cauth: bool

    @return: Dictionary of headers.
    @rtype: dict
    """

    # csrftoken is simply a 20 char random string.
    csrftoken = random_string(20)

    # Now make a call to the authenticator url.
    csrf2cookie = "csrf2_token_%s" % random_string(8)
    csrf2token = random_string(24)
    cookie = "csrftoken=%s; %s=%s" % (csrftoken, csrf2cookie, csrf2token)

    if include_cauth:
        CAUTH = session.cookies.get("CAUTH")
        cookie = "CAUTH=%s; %s" % (CAUTH, cookie)

    logging.debug("Forging cookie header: %s.", cookie)
    headers = {
        "Cookie": cookie,
        "X-CSRFToken": csrftoken,
        "X-CSRF2-Cookie": csrf2cookie,
        "X-CSRF2-Token": csrf2token,
    }

    return headers


def login(session, username, password, class_name=None):
    """
    Login on coursera.org with the given credentials.

    This adds the following cookies to the session:
        sessionid, maestro_login, maestro_login_flag
    """

    logging.debug("Initiating login.")
    try:
        session.cookies.clear(".coursera.org")
        logging.debug("Cleared .coursera.org cookies.")
    except KeyError:
        logging.debug("There were no .coursera.org cookies to be cleared.")

    # Hit class url
    if class_name is not None:
        class_url = CLASS_URL.format(class_name=class_name)
        r = requests.get(class_url, allow_redirects=False)
        try:
            r.raise_for_status()
        except requests.exceptions.HTTPError as e:
            logging.error(e)
            raise ClassNotFound(class_name)

    headers = prepare_auth_headers(session, include_cauth=False)

    data = {"email": username, "password": password, "webrequest": "true"}

    # Auth API V3
    r = session.post(AUTH_URL_V3, data=data, headers=headers, allow_redirects=False)
    try:
        r.raise_for_status()

        # Some how the order of cookies parameters are important
        # for coursera!!!
        v = session.cookies.pop("CAUTH")
        session.cookies.set("CAUTH", v)
    except requests.exceptions.HTTPError as e:
        raise AuthenticationFailed("Cannot login on coursera.org: %s" % e)

    logging.info("Logged in on coursera.org.")


def down_the_wabbit_hole(session, class_name):
    """
    Authenticate on class.coursera.org
    """

    auth_redirector_url = AUTH_REDIRECT_URL.format(class_name=class_name)
    r = session.get(auth_redirector_url)

    logging.debug(
        "Following %s to authenticate on class.coursera.org.", auth_redirector_url
    )

    try:
        r.raise_for_status()
    except requests.exceptions.HTTPError as e:
        raise AuthenticationFailed("Cannot login on class.coursera.org: %s" % e)

    logging.debug('Exiting "deep" authentication.')


def get_authentication_cookies(session, class_name, username, password):
    """
    Get the necessary cookies to authenticate on class.coursera.org.

    To access the class pages we need two cookies on class.coursera.org:
        csrf_token, session
    """

    # First, check if we already have the .coursera.org cookies.
    if session.cookies.get("CAUTH", domain=".coursera.org"):
        logging.debug("Already logged in on accounts.coursera.org.")
    else:
        login(session, username, password, class_name=class_name)

    try:
        session.cookies.clear("class.coursera.org", "/" + class_name)
    except KeyError:
        pass

    down_the_wabbit_hole(session, class_name)

    enough = do_we_have_enough_cookies(session.cookies, class_name)

    if not enough:
        raise AuthenticationFailed("Did not find necessary cookies.")

    logging.info("Found authentication cookies.")


def do_we_have_enough_cookies(cj, class_name):
    """
    Check whether we have all the required cookies
    to authenticate on class.coursera.org.
    """
    domain = "class.coursera.org"
    path = "/" + class_name

    return cj.get("csrf_token", domain=domain, path=path) is not None


def validate_cookies(session, class_name):
    """
    Checks whether we have all the required cookies
    to authenticate on class.coursera.org. Also check for and remove
    stale session.
    """
    if not do_we_have_enough_cookies(session.cookies, class_name):
        return False

    url = CLASS_URL.format(class_name=class_name) + "/class"
    r = session.head(url, allow_redirects=False)

    if r.status_code == 200:
        return True
    else:
        logging.debug("Stale session.")
        try:
            session.cookies.clear(".coursera.org")
        except KeyError:
            pass
        return False


def make_cookie_values(cj, class_name):
    """
    Makes a string of cookie keys and values.
    Can be used to set a Cookie header.
    """
    path = "/" + class_name

    cookies = [
        c.name + "=" + c.value
        for c in cj
        if c.domain == "class.coursera.org" and c.path == path
    ]

    return "; ".join(cookies)


def find_cookies_for_class(cookies_file, class_name):
    """
    Return a RequestsCookieJar containing the cookies for
    .coursera.org and class.coursera.org found in the given cookies_file.
    """

    path = "/" + class_name

    def cookies_filter(c):
        return c.domain == ".coursera.org" or (
            c.domain == "class.coursera.org" and c.path == path
        )

    cj = get_cookie_jar(cookies_file)

    new_cj = requests.cookies.RequestsCookieJar()
    for c in filter(cookies_filter, cj):
        new_cj.set_cookie(c)

    return new_cj


def load_cookies_file(cookies_file):
    """
    Load cookies file.

    We pre-pend the file with the special Netscape header because the cookie
    loader is very particular about this string.
    """

    logging.debug("Loading cookie file %s into memory.", cookies_file)

    cookies = StringIO()
    cookies.write("# Netscape HTTP Cookie File")
    cookies.write(open(cookies_file, "rU").read())
    cookies.flush()
    cookies.seek(0)
    return cookies


def get_cookie_jar(cookies_file):
    cj = cookielib.MozillaCookieJar()
    cookies = load_cookies_file(cookies_file)

    # nasty hack: cj.load() requires a filename not a file, but if I use
    # stringio, that file doesn't exist. I used NamedTemporaryFile before,
    # but encountered problems on Windows.
    cj._really_load(cookies, "StringIO.cookies", False, False)

    return cj


def get_cookies_cache_path(username):
    return os.path.join(PATH_COOKIES, username + ".txt")


def get_cookies_from_cache(username):
    """
    Returns a RequestsCookieJar containing the cached cookies for the given
    user.
    """

    logging.debug("Trying to get cookies from the cache.")

    path = get_cookies_cache_path(username)
    cj = requests.cookies.RequestsCookieJar()
    try:
        cached_cj = get_cookie_jar(path)
        for cookie in cached_cj:
            cj.set_cookie(cookie)
        logging.debug("Loaded cookies from %s", get_cookies_cache_path(username))
    except IOError:
        logging.debug("Could not load cookies from the cache.")

    return cj


def write_cookies_to_cache(cj, username):
    """
    Save RequestsCookieJar to disk in Mozilla's cookies.txt file format.

    This prevents us from repeated authentications on the
    accounts.coursera.org and class.coursera.org/class_name sites.
    """
    mkdir_p(PATH_COOKIES, 0o700)
    path = get_cookies_cache_path(username)
    cached_cj = cookielib.MozillaCookieJar()
    for cookie in cj:
        cached_cj.set_cookie(cookie)
    cached_cj.save(path)


def get_cookies_for_class(
    session, class_name, cookies_file=None, username=None, password=None
):
    """
    Get the cookies for the given class.

    We do not validate the cookies if they are loaded from a cookies file
    because this is intended for debugging purposes or if the coursera
    authentication process has changed.
    """
    if cookies_file:
        cookies = find_cookies_for_class(cookies_file, class_name)
        session.cookies.update(cookies)
        logging.info("Loaded cookies from %s", cookies_file)
    else:
        cookies = get_cookies_from_cache(username)
        session.cookies.update(cookies)
        if validate_cookies(session, class_name):
            logging.info("Already authenticated.")
        else:
            get_authentication_cookies(session, class_name, username, password)
            write_cookies_to_cache(session.cookies, username)


class TLSAdapter(HTTPAdapter):
    """
    A customized HTTP Adapter which uses TLS v1.2 for encrypted
    connections.
    """

    def init_poolmanager(self, connections, maxsize, block=False):
        self.poolmanager = PoolManager(
            num_pools=connections,
            maxsize=maxsize,
            block=block,
            ssl_version=ssl.PROTOCOL_TLSv1_2,
        )
